modules/collection/desktops/niri.nix

```diff
@@ -0,0 +1,222 @@
+{
+  pkgs,
+  config,
+  lib,
+  ...
+}: let
+  inherit (builtins) concatStringsSep;
+  inherit (lib.attrsets) mapAttrsToList;
+  inherit (lib.lists) singleton;
+  inherit (lib.modules) mkIf;
+  inherit (lib.options) mkEnableOption mkOption literalExpression;
+  inherit (lib.strings) concatMapStringsSep;
+  inherit (lib.trivial) pipe;
+  inherit (lib.types) listOf path attrsOf anything str lines submodule;
+
+  toNiriSpawn = commands:
+    concatMapStringsSep " " (arg: "\"${arg}\"") commands;
+
+  toNiriBinds = binds:
+    concatStringsSep "\n" (mapAttrsToList (
+        n: v: let
+          params = pipe v.params [
+            (mapAttrsToList (n': v': "${n'}=${v'}"))
+            (concatStringsSep " ")
+          ];
+          action =
+            if isNull v.spawn
+            then v.action
+            else toNiriSpawn v.spawn;
+        in "${n} ${params} {${action};}"
+      )
+      binds);
+
+  toNiriSpawnAtStartup = spawn:
+    concatMapStringsSep "\n" (
+      commands:
+        "spawn-at-startup " + (toNiriSpawn commands)
+    )
+    spawn;
+
+  niriEnvironment = let
+    variables = pipe (config.environment.sessionVariables // cfg.extraVariables) [
+      (mapAttrsToList (n: v: n + " \"${v}\""))
+      (concatStringsSep "\n")
+    ];
+  in ''
+    environment {
+      ${variables}
+    }
+  '';
+
+  bindsModule = submodule {
+    options = {
+      spawn = mkOption {
+        type = str;
+        default = null;
+        example = ["foot" "-e" "fish"];
+        description = ''
+          [niri's wiki]: https://github.com/YaLTeR/niri/wiki/Configuration:-Key-Bindings
+
+          The spawn action to run on button-press. For other actions, please see
+          {option}`binds.<keybind>.actions`. See [niri's wiki] for more information.
+        '';
+      };
+      action = mkOption {
```

Comment:
Could you add an assertion, in case one were to set both `spawn` and `action` at the same time? 

I believe you *should* be able to add assertions inside of the submodule itself, as to avoid mapping through the attribute set at Niri module level (since `lib.types.submodule` is a NixOS module itself). You can simply do this to access config:
```nix
submodule ({
  config,
  options,
  ...
}:
 ```

An example of this in the wild can be found in [upstream Hjem](https://github.com/feel-co/hjem/blob/ee5c671eeba5cddd94f2c7de3c36f019fae5854e/modules/common/user.nix#L22-L29).